ipc/src/uds.rs

```diff
@@ -0,0 +1,287 @@
+		let endpoint_addr = path.to_owned();
+		let meta_extractor = self.meta_extractor.clone();
+		let (stop_signal, stop_receiver) = oneshot::channel();
+		let (start_signal, start_receiver) = oneshot::channel();
+
+		remote.remote().spawn(move |handle| {
+			use self::tokio_uds::UnixListener;
+			use stream_codec::StreamCodec;
+
+			let listener = match UnixListener::bind(&endpoint_addr, handle) {
+				Ok(l) => l,
+				Err(e) => {
+					start_signal.complete(Err(e));
+					return future::ok(()).boxed();
+				}
+			};
+
+			start_signal.complete(Ok(()));
+			let remote = handle.remote().clone();
+			let connections = listener.incoming();
+
+			let server = connections.for_each(move |(unix_stream, client_addr)| {
+				trace!("Accepted incoming UDS connection: {:?}", client_addr);
+
+				let meta = meta_extractor.extract(&RequestContext { endpoint_addr: &client_addr });
+				let service = Service::new(rpc_handler.clone(), meta);
+				let (writer, reader) = unix_stream.framed(StreamCodec).split();
+				let responses = reader.and_then(
+					move |req| service.call(req).then(|response| match response {
+						Err(e) => {
+							warn!(target: "uds", "Error while processing request: {:?}", e);
+							future::ok(String::new())
+						},
+						Ok(None) => {
+							trace!(target: "uds", "JSON RPC request produced no response");
+							future::ok(String::new())
+						},
+						Ok(Some(response_data)) => {
+							trace!(target: "uds", "Sent response: {}", &response_data);
+							future::ok(response_data)
+						}
+					})
+				);	
+
+				let writer = writer.send_all(responses).then(move |_| {
+					trace!(target: "uds", "Peer {:?}: service finished", client_addr);
+					Ok(())
+				});
+
+				remote.spawn(|_| writer);
+
+				Ok(())
+			});
+
+			let stop = stop_receiver.map_err(|_| std::io::ErrorKind::Interrupted.into());
+			server.select(stop).map(|_| ()).map_err(|_| ()).boxed()
+		});
+
+		match start_receiver.wait().expect("Message should always be sent") {
+			Ok(()) => Ok(Server { path: path.to_owned(), remote: Some(remote), stop: Some(stop_signal) }),
+			Err(e) => Err(e)
+		}	
+	}	
+}
+
+
+/// IPC Server handle
+pub struct Server {
+	path: String,
+	remote: Option<reactor::Remote>,
+	stop: Option<oneshot::Sender<()>>,
+}
+
+impl Server {
+	/// Closes the server (waits for finish)
+	pub fn close(mut self) {
+		self.stop.take().unwrap().complete(());
+		self.remote.take().unwrap().close();
+		self.clear_file();
+	}
+
+	/// Wait for the server to finish
+	pub fn wait(mut self) {
+		self.remote.take().unwrap().wait();
+	}
+
+	/// Remove socket file
+	fn clear_file(&self) {
+		::std::fs::remove_file(&self.path).unwrap_or_else(|_| {}); // ignore error, file could have been gone somewhere
+	}
+}
+
+impl Drop for Server {
+	fn drop(&mut self) {
+		self.stop.take().map(|stop| stop.complete(()));
+		self.remote.take().map(|remote| remote.close());	
+		self.clear_file();	
+	}
+}
+
+pub fn server<I, M: Metadata, S: Middleware<M> + Send + Sync + 'static>(
```

Comment:
`Send + Sync + 'static` bounds are not necessary here, they are part of `Middleware` trait: https://github.com/ethcore/jsonrpc/blob/b36a57c4bf449c431dc59f0e88236026eda62ea7/core/src/middleware.rs#L8